### PR TITLE
feat: add upstream lineage dependencies for view/materialized view in bq extractor

### LINKED ISSUE #36: Building upstream dependencies for View / Materilized View type in Bigquery extractor
**Problem Statement**

To build lineage for View/Materialized View, We need to extract upstream dependencies for such table type.Currently, there is no way to find out upstream dependencies for View/Materialized views in bigquery extractor.

**Solution**

We can have configurable option `build_view_lineage` in bigquery recipe which is when `true` checks for VIEW or MATERIALIZED_VIEW type, fetches Query from it's metadata and hence extracts upstream dependencies by parsing SQL query. This information can be added in `lineage.upstreams` in built asset.

### PR DESCRIPTION
**Pull request by utsav14nov**

### CODE REVIEW

File: plugins/extractors/bigquery/bigquery.go [hunk 1]
```diff
@@ -432,15 +435,25 @@ func (e *Extractor) buildAsset(ctx context.Context, t *bigquery.Table, md *bigqu
...
 		Name:        t.TableID,
 		Type:        "table",
 		Description: md.Description,
 		Service:     "bigquery",
 		Data:        table,
 		Labels:      md.Labels,
-	}, nil
+	}
+
+	if e.config.BuildViewLineage == true && (md.Type == bigquery.ViewTable || md.Type == bigquery.MaterializedView) {
```
Comment (bsushmith) [inline plugins/extractors/bigquery/bigquery.go]: Can you fix the lint errors

File: plugins/extractors/bigquery/upstream/parser.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
+package upstream
+
+import (
+	"regexp"
+	"strings"
+)
+
+type QueryParser func(query string) []Resource
```
Comment (bsushmith) [inline plugins/extractors/bigquery/upstream/parser.go:8]: can you add some test cases for the parsing logic?

File: plugins/extractors/bigquery/bigquery.go [hunk 3]
```diff
@@ -432,15 +435,25 @@ func (e *Extractor) buildAsset(ctx context.Context, t *bigquery.Table, md *bigqu
...
+
+	if e.config.BuildViewLineage && (md.Type == bigquery.ViewTable || md.Type == bigquery.MaterializedView) {
+		query := getViewQuery(md)
+		upstreamResources := getUpstreamResources(query)
+		asset.Lineage = &v1beta2.Lineage{
+			Upstreams: upstreamResources,
+		}
+	}
+
+	return asset, nil
```
Comment (haveiss) [inline plugins/extractors/bigquery/bigquery.go:456]: would it possible we cover this in test?

File: plugins/extractors/bigquery/bigquery_test.go [hunk 4]
```diff
@@ -229,6 +229,37 @@ func TestExtract(t *testing.T) {
...
+				return func(max int64) int64 {
+					return r.Int63n(max)
+				}
+			}
+		}
+
+		t.Run("should return upstream lineage dependencies of type view", func(t *testing.T) {
+			actual := runTest(t, cfg, randFn(1))
+
+			utils.AssertJSONFile(t, "testdata/expected-assets-mixed.json", actual, jsondiff.FullMatch)
```
Comment (haveiss) [inline plugins/extractors/bigquery/bigquery_test.go]: the test is failing, the testdata should be `testdata/expected-assets-view-lineage.json` right?

Comment (haveiss) [reply plugins/extractors/bigquery/bigquery_test.go]: it's still failing

File: plugins/extractors/bigquery/testdata/data.yaml [hunk 5]
```diff
@@ -66,6 +66,22 @@ projects:
...
+              type: STRING
+              mode: REQUIRED
+          data:
+            - id: 1
+              name: alice
+            - id: 2
+              name: bob
+          metadata:
+            - type: VIEW
+              view_query: select id,name from test-project-id.dataset1.table_a
```
Comment (haveiss) [inline plugins/extractors/bigquery/testdata/data.yaml]: i don't think it is supported by the emulator, 
currently it is hardcoded to `TABLE` [here](https://github.com/goccy/bigquery-emulator/blob/8ccde288d63846122e085433e0a03edfadba361c/types/types.go#L35-L57), we can skip this test for now if it's not feasible

Comment (utsav14nov) [reply plugins/extractors/bigquery/testdata/data.yaml]: @haveiss Yes, as seen VIEW is not supported in table metadata in BQ emulator, hence not able to test this through current flow.
I'll update the code

File: plugins/extractors/bigquery/bigquery.go (hunk 3 above)
Comment (utsav14nov) [reply plugins/extractors/bigquery/bigquery.go:456]: As discussed, this is not possible through current test setup as BQ emulator doesn't support VIEW

**Review by haveiss — APPROVED** [review]

**Pull request merged**